### ISSUE TITLE
Append implementation of Vec doesn't look safe

### ISSUE THREAD

**Issue by Stargateur** [rust-lang/rust#64107]
Randomly looking at the code of `append()`, https://github.com/rust-lang/rust/blob/master/src/liballoc/vec.rs#L1157, doesn't look safe:

```rust
self.append_elements(other.as_slice() as _);
other.set_len(0);
```

As, we already copy and set the new `len` before reset other vector `len`, I think we could have a race in a panic situation.

Should we not use a guard or use maybeunit somehow ?

**Comment by Centril** [comment]
cc @RalfJung @Gankra

**Comment by nagisa** [comment]
Looks fine to me. Here’s the sequence of operations reproduced:

```rust
// won't unwind
let count = (*other).len();
// could unwind, but nothing changes, data remains valid
self.reserve(count);
// won't unwind
let len = self.len();
// won't unwind
ptr::copy_nonoverlapping(other as *const T, self.as_mut_ptr().add(len), count); 
// could unwind (panic on overflow), but structures remain well formed, 
// as the data copied from the other vector is not considered "used"
self.len += count;
// won't unwind, though this could be changed to `other.len = 0` to make sure of it.
// Alternatively this could be swapped with the statement above for more confidence.
other.set_len(0);
```

**Comment by Stargateur** [comment]
@nagisa the problem would be similar to the one describe here, https://github.com/rust-lang/rust/blob/5a56e05abd34e1936df74625c1f40cb6fee0cd4a/src/libstd/io/mod.rs#L310. If we did this for this case I don't see why we should not for the `append()` case

**Comment by jakubadamw** [comment]
> Between self.len += count; and other.set_len(0); you could have a panic that will make both vector drop the same content.

How so? if `self.len += count` overflows and unwinds then `self.len` wouldn't have changed so the additional `count` elements will not be dropped when `self` is dropped.

**Comment by Gankra** [comment]
The guard in append_to_string is trying to cope with the fact that we're calling some opaque generic function, which may push invalid utf8 and then panic. Because the `len` of the String is the `len` of the Vec, we cannot simply "not" update the String's length on errors or panics -- it has already been updated by the opaque function. 

To solve this, the guard remembers the old length and forces it back to the old value if anything goes wrong.

Vec::append on the other hand does not invoke arbitrary generic code in the middle of its work. It calls a series of trusted functions (other Vec methods and core ptr methods which are straight up UB if you violate their contracts).

The language does not "spuriously" panic, it only panics if you do something that can panic. This is critical for unsafe code to be tractable to write. set_len cannot panic.

**Comment by Gankra** [comment]
A minor addendum to pre-empt this concern:

set_len *can* panic, because it has a debug_assert, but:

* `debug_assert`s in the stdlib essentially don't exist, because outside of our own CI everyone only has release builds of the stdlib

* This assert is merely attempting to make it easier to find UB early in testing. set_len is unsafe and declares this debug assert as a safety constraint. So we have permission to cause UB if you trip it.

**Comment by Gankra** [comment]
A second addendum to also pre-empt this concern: append_to_string does not actually sufficiently guard against misbehaving opaque functions. If one scribbled over the existing bytes, it would lead to UB, and we don't check for this.

However append_to_string is not a public API, it's just a helper inside this unsafe module. It is reasonably assuming the set of implementations it controls only append, but still correctly handling the fact that they can and do panic.

**Comment by Centril** [comment]
@Gankra This all seems like great information to add to the code; perhaps you could write an internal docs PR?

**Comment by Gankra** [comment]
append_to_string already has excellent documentation. I skipped over the docs when I wrote my comments above, but looking at them now, they cover the issue quite well. They could possibly spell out things slightly more explicitly, but the presence of a generic callback in unsafe code is a pretty standard "wait think about exception-safety" red flag.

It also doesn't make sense to add any comments to Vec::append discussing how exception-safety *isn't* a concern, because it's not at all notable in this regard. (Life is too short to write out everything which *isn't* a concern for every single function). Anyone spending time in libcollections (or really writing unsafe code) needs to concern themselves with understanding exception-safety.

Furthermore, I already carefully documented the problem [in the Rustonomicon section on exception-safety](https://doc.rust-lang.org/nomicon/exception-safety.html).

I merely wrote out the comments here to carefully explain the issue to what is presumably an enthusiastic new unsafe rust programmer, and to clarify the confusion to anyone passing by this issue.

That said, I believe this issue can now be closed.
